magento/magento2
# magento/magento2#37796: Fixed issue with adding new country regions

Fixed issue when the configuration 'general/region/state_required' gets reset to the default countries with a required states.

### Description (*)
According to [#37796 ](https://github.com/magento/magento2/issues/37796) in \Magento\Directory\Setup\DataInstaller::class weren't declared proper logic of saving configuration 'general/region/state_required'. That could lead to problems for the web-site if you changed that configuration from the admin panel previously and then decided to add some new regions for some specific country. 
Based on that I've fixed the bug with the help of getting current configuration value before any DB updates, then I'm merging it with new data that should be added and finally updating the configuration with proper value.

### Fixed Issues (if relevant)
1. Fixes magento/magento2#37796

### Manual testing scenarios (*)
1. Go to the admin panel and open Stores > Settings > Configuration > General > General > State Options
2. Unselect US from "State is Required for" and save configuration.
![image](https://github.com/magento/magento2/assets/75758543/f00f6161-00af-40a8-9073-53ad1315daf6)
3.Then go to the project root and check with CLI command config value and make sure that US is not in there.
![image](https://github.com/magento/magento2/assets/75758543/e323c1f7-2fd6-44e1-9026-84135271524d)
4.Then use this custom module to add Japan with regions to state_required config.
[RequiredVendor.zip](https://github.com/magento/magento2/files/12894168/RequiredVendor.zip)
5.In case if my fix is applied you would get a proper result for config - including JP but still excluding US.


### Contribution checklist (*)
